github/ratelimit_test.go

```diff
@@ -0,0 +1,331 @@
+	"/network_error": response{
+		statusCode: http.StatusTeapot,
+		headers:    nil,
+		response:   apiResponse{},
+		err:        fmt.Errorf("network error"),
+	},
+}
+
+type response struct {
+	statusCode int
+	headers    func(time.Time) map[string]string
+	response   apiResponse
+	err        error
+}
+
+type apiResponse struct {
+	apiErrorResponse
+	Data string `json:"data,omitempty"`
+}
+
+func rateLimitHeaders(wait time.Duration) func(time.Time) map[string]string {
+	return func(when time.Time) map[string]string {
+		return map[string]string{
+			"X-RateLimit-Reset":     strconv.FormatInt(when.Add(wait).Unix(), 10),
+			"X-RateLimit-Remaining": "0",
+		}
+	}
+}
+
+func abuseHeaders(wait time.Duration) func(time.Time) map[string]string {
+	return func(when time.Time) map[string]string {
+		return map[string]string{
+			"Retry-After": strconv.FormatInt(int64(wait.Seconds()), 10),
+		}
+	}
+}
+
+func newResponse(content apiResponse, headers func(time.Time) map[string]string, statusCode int) (*http.Response, error) {
+	data, err := json.Marshal(content)
+	if err != nil {
+		return nil, err
+	}
+
+	respWriter := httptest.NewRecorder()
+	respWriter.WriteHeader(statusCode)
+	_, err = respWriter.Write(data)
+	if err != nil {
+		return nil, err
+	}
+
+	response := respWriter.Result()
+
+	if headers != nil {
+		for k, v := range headers(time.Now()) {
+			response.Header.Set(k, v)
+		}
+	}
+
+	return response, err
+}
+
+func newRequest(url string) *http.Request {
+	req, _ := http.NewRequest("GET", url, http.NoBody)
+	return req
+}
+
+// gitHubTransportResponseMock avoids calling the network to get the response directly from testcases
+type gitHubTransportResponseMock struct{}
+
+func (gh *gitHubTransportResponseMock) RoundTrip(req *http.Request) (*http.Response, error) {
+	url := req.URL.String()
+	tc, ok := testCases[url]
+	if !ok {
+		return newResponse(apiResponse{}, nil, http.StatusOK)
+	}
+
+	if tc.err != nil {
+		return &http.Response{}, tc.err
+	}
+
+	return newResponse(tc.response, tc.headers, tc.statusCode)
+}
+
+func TestRateLimitSuite(t *testing.T) {
+	suite.Run(t, new(RateLimitSuite))
+}
+
+type RateLimitSuite struct {
+	suite.Suite
+	transport  *RateLimitTransport
+	loggerMock *testutils.LoggerMock
+}
+
+func (s *RateLimitSuite) SetupSuite() {
+	defaultAbuseRetryAfter = default403Reset
+}
+
+func (s *RateLimitSuite) SetupTest() {
+	s.loggerMock = &testutils.LoggerMock{}
+	s.transport = NewRateLimitTransport(&gitHubTransportResponseMock{}, s.loggerMock)
+}
+
+func (s *RateLimitSuite) TearDownSuite() {
+}
+
+// TestNoLimit ensures that the regular case: no rate limit nor abuse, is not considered a RateLimit case
+func (s *RateLimitSuite) TestNoLimit() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/normal"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Equal("", s.loggerMock.Next())
+	s.NoError(err)
```

Comment:
error should be checked before anything else as prev checks can weirdly fail because the code took error path but it won't be reported by the test runner. (or will be reported _after_ other fails and can easily be missed)

Reply:
done by b43b7d2